### PR TITLE
Add Mixtral model for MoE demo

### PR DESCRIPTION
**Pull request by pstjohn**
### Summary                                                                                

Adds a Mixtral Mixture-of-Experts model to bionemo-recipes/models/mixtral/ using TransformerEngine, following the same pattern as the existing llama3
model.

### MoE Implementation

The core MoE block (NVMixtralSparseMoeBlock) uses:
- te.GroupedLinear for efficient parallel expert FFN computation (gate_up + down projections)
- te.moe_permute / te.moe_unpermute with map_type="index" for token-to-expert routing
- Standard nn.Linear router (kept in bf16 during FP8 training via te.autocast(enabled=False))
- SwiGLU activation with fused gate/up projection split

Weight conversion handles the structural difference between HF's stacked 3D expert tensors ([num_experts, out, in]) and TE's per-expert GroupedLinear
weights (weight0, weight1, ...).

### Base test class improvements

- Added clear_gpu_memory fixture (gc + cuda cache clear before/after each test) to prevent OOM cascading
- get_converted_te_model_checkpoint now frees the HF model and moves TE to CPU before saving (save_pretrained clones state dict internally)
- test_golden_values and test_golden_values_thd now run models sequentially to support large models

## Summary by CodeRabbit

* **New Features**
  * Added support for Mixtral mixture-of-experts model with Transformer Engine backend.
  * Improved token packing and sequence batching with overflow handling and sample splitting.

* **Bug Fixes**
  * Fixed typo in error logging messages.
  * Enhanced validation for tensor format handling in data collators.

* **Tests**
  * Added comprehensive test suite for Mixtral model conversions and FP8 quantization.
  * Improved GPU capability detection and pattern-based model configuration handling.

### CODE REVIEW

File: bionemo-recipes/models/mixtral/tests/test_modeling_mixtral.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            cp_loss_atol=0.5,
+            cp_loss_rtol=0.25,
+        )
+
+    @pytest.mark.skip(
+        reason="MoE routing is batch-dependent: padding tokens in BSHD affect softmax normalization "
+        "and top-k expert selection, so BSHD and THD produce fundamentally different routing decisions."
+    )
+    def test_golden_values_thd(self, te_attn_backend):
+        """Skip: BSHD vs THD comparison is not meaningful for MoE models."""
```
Comment (pstjohn) [inline bionemo-recipes/models/mixtral/tests/test_modeling_mixtral.py]: is this correct?

**Comment by pstjohn** [comment]
@coderabbitai resolve

**Comment by jomitchellnv** [comment]
@coderabbitai run

File: bionemo-recipes/models/mixtral/tests/test_modeling_mixtral.py (hunk 1 above)
Comment (pstjohn) [reply bionemo-recipes/models/mixtral/tests/test_modeling_mixtral.py]: it wasnt -- turns out this was due to 

* the attention mask being calculated incorrectly (also a bug with llama3)
* left-side padding as default in the mixtral tokenizer, which we don't support

File: bionemo-recipes/models/llama3/tests/test_cp_bshd.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ def get_te_model_checkpoint(tmp_path):
     """
     # Use the 1B model for practical testing (8B model requires too much memory)
     model_hf = AutoModelForCausalLM.from_pretrained(
-        "meta-llama/Llama-3.2-1B-Instruct", revision="c731040f", dtype=torch.bfloat16
+        "meta-llama/Llama-3.2-1B-Instruct", revision="9213176", dtype=torch.bfloat16
```
Comment (jomitchellnv) [inline bionemo-recipes/models/llama3/tests/test_cp_bshd.py:94]: I was actually plumbing this down through configs. `dataset.tokenizer_revision`

File: bionemo-recipes/models/mixtral/tests/common/__init__.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+- TestTolerances: Dataclass for model-specific numerical tolerances
+- Distributed testing utilities for multi-GPU tests
+- Shared fixtures for common test requirements
+
+Example usage:
+
+    ```python
+    from tests.common import BioNeMoModelTester, BioNeMoModelTest, TestTolerances
+
+    class ESM2ModelTester(BioNeMoModelTester):
```
Comment (jomitchellnv) [inline bionemo-recipes/models/mixtral/tests/common/__init__.py:32]: not an ESM2 model

File: bionemo-recipes/models/mixtral/tests/common/fixtures.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    yield
+    os.environ.pop("NVTE_DEBUG", None)
+
+
+ALL_RECIPES = [
+    recipe_module.DelayedScaling(),
+    recipe_module.Float8CurrentScaling(),
+    recipe_module.Float8BlockScaling(),
+    recipe_module.MXFP8BlockScaling(),
+    recipe_module.NVFP4BlockScaling(disable_rht=True, disable_stochastic_rounding=True),
```
Comment (jomitchellnv) [inline bionemo-recipes/models/mixtral/tests/common/fixtures.py:59]: why disable RHT and stochastic rounding

Comment (jomitchellnv) [reply bionemo-recipes/models/mixtral/tests/common/fixtures.py:59]: just for determinism right?

File: bionemo-recipes/models/mixtral/.ruff.toml [hunk 5]
```diff
@@ -0,0 +1 @@
+extend = "../.ruff.toml"
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply bionemo-recipes/models/mixtral/.ruff.toml:1]: watching

File: bionemo-recipes/models/mixtral/tests/common/fixtures.py (hunk 4 above)
Comment (pstjohn) [reply bionemo-recipes/models/mixtral/tests/common/fixtures.py:59]: i actually think this was to get it to run on my 5090 locally 🤷 . I could probably edit these to xfail / etc. better depending on hardware, but let's do that in another PR. this is just copying those to this directory

File: bionemo-recipes/models/mixtral/tests/common/__init__.py (hunk 3 above)
Comment (pstjohn) [reply bionemo-recipes/models/mixtral/tests/common/__init__.py:32]: right, this is a docstring showing _how_ to use the common test library. but these are copied from the esm2 model folderr

File: bionemo-recipes/models/llama3/tests/test_cp_bshd.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/models/llama3/tests/test_cp_bshd.py:94]: i think this was just a typo, but yeah, not a bad idea

File: bionemo-recipes/models/mixtral/.ruff.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pstjohn) [reply bionemo-recipes/models/mixtral/.ruff.toml:1]: bad bot

File: bionemo-recipes/models/mixtral/tests/common/__init__.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        def get_model_class(self):
+            return NVEsmForMaskedLM
+        # ... implement other abstract methods
+    ```
+"""
```
[the remark this answers is not in the document]
Comment (pstjohn) [reply bionemo-recipes/models/mixtral/tests/common/__init__.py:37]: noted but let me do that elsewhere so i don't trigger CI on every model
